xds/src/main/java/io/grpc/xds/RingHashLoadBalancer.java

```diff
@@ -343,35 +345,60 @@ public PickResult pickSubchannel(PickSubchannelArgs args) {
         }
       }
 
-      // Try finding the first non-TRANSIENT_FAILURE connection.
-      int attempt = 0;
-      Subchannel subchannel;
-      ConnectivityStateInfo stateInfo;
-      do {
-        int index = (mid + attempt) % ring.size();
+      // Try finding a READY subchannel. Starting from the ring entry next to the RPC's hash.
+      // If the one of the first two subchannels is not in TRANSIENT_FAILURE, return result
+      // based on that subchaannel. Otherwise, fail the pick unless a READY subchannel is found.
+      // Meanwhile, trigger connection for the first subchannel that is in IDLE if no subchannel
+      // before it is in CONNECTING or READY.
+      boolean hasPending = false;
+      Set<EquivalentAddressGroup> attemptedAddrs = new HashSet<>();
+      for (int i = 0; i < ring.size(); i++) {
+        int index = (mid + i) % ring.size();
         EquivalentAddressGroup addrKey = ring.get(index).addrKey;
-        subchannel = pickableSubchannels.get(addrKey);
-        stateInfo = getSubchannelStateInfoRef(subchannel).get();
-        attempt++;
-      } while (attempt < ring.size() && stateInfo.getState() == TRANSIENT_FAILURE);
-
-      if (stateInfo.getState() == READY) {
-        return PickResult.withSubchannel(subchannel);
-      }
-      if (stateInfo.getState() == TRANSIENT_FAILURE) {  // all connections in TRANSIENT_FAILRE
-        // TODO(chengyuanzhang): shall we use some other error?
-        return PickResult.withError(stateInfo.getStatus());
-      }
-      if (stateInfo.getState() == IDLE) {
-        final Subchannel finalSubchannel = subchannel;
-        syncContext.execute(new Runnable() {
-          @Override
-          public void run() {
-            finalSubchannel.requestConnection();
+        if (attemptedAddrs.contains(addrKey)) {
+          continue;
+        }
+        attemptedAddrs.add(addrKey);
+        SubchannelView subchannel = pickableSubchannels.get(addrKey);
+        if (subchannel.stateInfo.getState() == READY) {
+          return PickResult.withSubchannel(subchannel.subchannel);
+        }
+        if (subchannel.stateInfo.getState() == TRANSIENT_FAILURE) {
+          continue;
+        }
+        if (!hasPending) {  // first non-failing subchannel
+          if (subchannel.stateInfo.getState() == IDLE) {
+            final Subchannel finalSubchannel = subchannel.subchannel;
+            syncContext.execute(new Runnable() {
+              @Override
+              public void run() {
+                finalSubchannel.requestConnection();
```

Comment:
Only request connection for subchannels in IDLE. Btw, request connection for subchannels in any other state is just a no-op.

Reply:
I'm in `syncContext`, should check again to make sure not in `SHUTDOWN` state.

Reply:
Hmm... That should be fine, it will just be a no-op if it already exited IDLE. It's less verbose to not do that. But anyway, I am fine with either.

https://github.com/grpc/grpc-java/blob/4896a0fee87789b0f8e844255356bef3a68010a8/core/src/main/java/io/grpc/internal/PickFirstLoadBalancer.java#L158-L164